smart_contract/executor/src/tokens.rs

```diff
@@ -0,0 +1,478 @@
+                    $token_name => {
+                        let token = <$token_ty>::try_from($token.clone()).dbg_expect("Internal error");
+                        $callback!(token);
+                    } )+
+                    _ => {}
+                }
+
+            };
+        }
+
+        macro_rules! map_token_type {
+            ($callback:ident) => { $(
+                $callback!($token_ty); )+
+            };
+        }
+
+        pub(crate) use map_token;
+        pub(crate) use map_token_type;
+    };
+}
+
+macro_rules! token {
+    ($($meta:meta)* $item:item) => {
+        #[derive(PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+        #[derive(Clone, iroha_executor_derive::Token)]
+        #[derive(iroha_schema::IntoSchema)]
+        $($meta)*
+        $item
+    };
+}
+
+declare_tokens! {
+    // peer
+    "CanUnregisterAnyPeer" => crate::tokens::peer::CanUnregisterAnyPeer,
+
+    // domain
+    "CanUnregisterDomain" => crate::tokens::domain::CanUnregisterDomain,
+    "CanSetKeyValueInDomain" => crate::tokens::domain::CanSetKeyValueInDomain,
+    "CanRemoveKeyValueInDomain" => crate::tokens::domain::CanRemoveKeyValueInDomain,
+
+    // account
+    "CanUnregisterAccount" => crate::tokens::account::CanUnregisterAccount,
+    "CanMintUserPublicKeys" => crate::tokens::account::CanMintUserPublicKeys,
+    "CanBurnUserPublicKeys" => crate::tokens::account::CanBurnUserPublicKeys,
+    "CanMintUserSignatureCheckConditions" => crate::tokens::account::CanMintUserSignatureCheckConditions,
+    "CanSetKeyValueInUserAccount" => crate::tokens::account::CanSetKeyValueInUserAccount,
+    "CanRemoveKeyValueInUserAccount" => crate::tokens::account::CanRemoveKeyValueInUserAccount,
+
+    // asset definition
+    "CanUnregisterAssetDefinition" => crate::tokens::asset_definition::CanUnregisterAssetDefinition,
+    "CanSetKeyValueInAssetDefinition" => crate::tokens::asset_definition::CanSetKeyValueInAssetDefinition,
+    "CanRemoveKeyValueInAssetDefinition" => crate::tokens::asset_definition::CanRemoveKeyValueInAssetDefinition,
+
+    // asset
+    "CanRegisterAssetsWithDefinition" => crate::tokens::asset::CanRegisterAssetsWithDefinition,
+    "CanUnregisterAssetsWithDefinition" => crate::tokens::asset::CanUnregisterAssetsWithDefinition,
+    "CanUnregisterUserAsset" => crate::tokens::asset::CanUnregisterUserAsset,
+    "CanBurnAssetsWithDefinition" => crate::tokens::asset::CanBurnAssetsWithDefinition,
+    "CanBurnUserAsset" => crate::tokens::asset::CanBurnUserAsset,
+    "CanMintAssetsWithDefinition" => crate::tokens::asset::CanMintAssetsWithDefinition,
+    "CanTransferAssetsWithDefinition" => crate::tokens::asset::CanTransferAssetsWithDefinition,
+    "CanTransferUserAsset" => crate::tokens::asset::CanTransferUserAsset,
+    "CanSetKeyValueInUserAsset" => crate::tokens::asset::CanSetKeyValueInUserAsset,
+    "CanRemoveKeyValueInUserAsset" => crate::tokens::asset::CanRemoveKeyValueInUserAsset,
```

Comment:
Looks like this repetition could be easily handled inside macro

Reply:
I don't know how

Reply:
The usage could be something like:
```rust
declare_tokens! {
    crate::tokens::asset::{
        CanRegisterAssetsWithDefinition,
        CanUnregisterAssetsWithDefinition,
        // ...
    };
    crate::tokens::account::{
        // ...
    };
    // ...
}
```

And for macro implementation you can use `stringify!`

Reply:
check out now, this is the best I was able to come up with

Reply:
I see, it's hard because you need to use interpolation inside inner macro